README.md
# whyem-etl

![Python Versions](https://img.shields.io/github/pipenv/locked/python-version/miyeung/whyem-etl)
[![CircleCI](https://circleci.com/gh/circleci/circleci-docs.svg?style=shield)](https://circleci.com/gh/miyeung/whyem-etl)
[![codecov](https://codecov.io/gh/miyeung/whyem-etl/branch/master/graph/badge.svg)](https://codecov.io/gh/miyeung/whyem-etl)
[![black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

## Description

whyem-etl is the pet project I use as a sandbox for experimentation and
learning. Therefore should not be used in production. It consists in a Python
web application that provides APIs to get various information from a dataset
containing job applications information.

The current APIs allows to:

- Enrich the data in an ETL fashion
- Get aggregated data.

In term of toolings, the project comes with:

- Application packaging (pipenv)
- Code quality checks
  - Linter (flake8)
  - Formatter (black)
  - Vulnerabilities in dependencies and code (pipenv / bandit)
  - Automated unit testing (pytest)
- Local deployment using docker and docker-compose
- Continuous Integration (CI) using CircleCI running the above code quality checks.

Provided APIs are:

| Type | Name               | Port | Description                                                                         |
| ---- | ------------------ | ---- | ----------------------------------------------------------------------------------- |
| POST | /jobsByContinent   | 5000 | Returns the aggregated jobs information by continent given the profession_ids.      |
| GET  | /consolidateData   | 5000 | Consolidates the jobs data by adding the Continent information in an ETL fashion.   |
| GET  | /checkDbConnection | 5000 | Check connectivity between web server database by returning first row of jobs data. |
| GET  | /                  | 5000 | Check connectivity with web server.                                                 |

## Getting started

### Prerequisites

This application requires:

- Python 3.7
- Docker 19.03.1 or higher
- Docker-compose 1.241 or higher

#### Python

```bash
# Check your current Python version
python --version # should be 3.7

# Update repo list
sudo apt-get update

# Add Deadsnake repo
sudo apt-get install software-properties-common
sudo add-apt-repository ppa:deadsnakes/ppa
sudo apt-get update

# Install Python 3.7
sudo apt-get install python3.7

# Check version
python --version # should be 3.7

# Install pip3
sudo apt-get install python3-pip
```

The project uses `pipenv` to handle Python packaging. We will need to install it
using pip.

```bash
# Check that you use pip3
pip --version # should use Python 3.7

# If not, change default pip to use pip3
ln -s /home/$USER/.local/bin/pip3 /home/$USER/.local/bin/pip

# Check that you now use pip3
pip --version # Should use Python 3.7

# Install pipenv
pip install pipenv --upgrade --user

# Check pipenv installation
pipenv --version # pipenv, version 2018.11.26
```

#### Docker and Compose

Follow instructions from official docs `https://docs.docker.com/engine/install/`
and `https://docs.docker.com/compose/install/`. Pay special attention to the
version your download for Compose. The current docker-compose.yml file is not
compatible with all Compose version.

You will also need to be run Docker with non-root access:
`https://docs.docker.com/engine/install/linux-postinstall/`.

### Installing project dependencies

Once the above tools installed, run the following commands. It will install
Python dependencies using pipenv.

```bash
# Init Python environment with pipenv
make init
```

### Run web application and database

To run the web application and its Postgres database, run the following:

```bash
# Build docker image
make docker-image

# Run web application and its database with docker-compose
make compose-up
```

It will create a docker image for the Python web application and pulls a Postgres
database image. Once generated and downloaded, they will be run as containers
within a virtual network, and connected altogether. Mocked data located in
`./data/dump.sql` will automatically be loaded in the database.

### Check connectivity with web application

```bash
curl http://127.0.0.1:5000/
# Hello, World!
```

### Check web application connectivity with database

```bash
curl http://127.0.0.1:5000/checkDbConnection
```

The response should be the first row in database:

```json
{
    "candidates":2,
    "job_reference":"00008062b54689de44d78b240fff3b89",
    "office_continent":"europe",
    "office_latitude":48.8669091,
    "office_longitude":2.3505013,
    "organization_reference":"1e9bd0d26870a3962298751941adc09a",
    "profession_id":15
}
```

### Add continent information to data

In the original data, the continent information is not provided. To tackle this
issue, an `office_continent` column has been added to the table and the
`/consolidateData` implemented.

`/consolidateData` works in an ETL fashion: it extracts the data from a
database, transforms it by adding the continent information and loads it back.
It can takes a few minutes to be run.

```bash
curl http://127.0.0.1:5000/consolidateData
```

This API is yet to be improved on how continent information is derived given
(latitude, longitude). As of today, the API:

- add continent labels for `EUROPE`, `ASIA` and `AMERICA`
- for rows that could not be labeled, `OTHERS` is added
- has a success rate of 98.23% (475 rows with `OTHERS` among 26873 total rows).

For data quality evaluation and future improvements, the API response contains
the success rate and the failed rows references. See example below:

```json
{
    "consolidation_success_rate": 98.23242659918877,
    "failed_rows": [
        "0069be0f45bf177a85fcb8300cece198",
        "00963dc1881f268931a39e517ddd7580",
        "00d70de33e4eb465d90c94758e22c2c8",
        "01268e220faeec583b1da2b21134b2d7",
        "012a0b12eda34452467fae4bdbe6ca9e",
        "023e96ef31fbfea30c6044b8997ece74",
        "02411b211404ee28ba6b203f10050822",
        "02706b187a03ca3a7f2780bcefb1ace2",
        "..."
    ]
}
```

### Get jobs information by continent

To get the total number of job offers aggregated by continent for a given list
of professions, we use the `/jobsByContinent` API. See below example usage for
profession_id 12 and 15.

```bash
# Getting information for profession_id = 12 and 15
curl --header "Content-Type: application/json" \
    --request POST \
    --data '{"profession_ids": [12,15]}' \
    http://127.0.0.1:5000/jobsByContinent
```

See below the response.

```json
{
    "AMERICA": 167,
    "ASIA": 5,
    "EUROPE": 35613,
    "OTHERS": 149
}
```

Note that the API provides also provides the number of job offers on
unidentified continents.

## Tooling

### Python application

```bash
# Set up pipenv project, dependencies and install pre-commit
make init

# Syntax check with Flake8 linter
make flake8

# Codestyle check with Black formatter
make black

# Security dependencies check with Pipenv
make check-deps

# Security code check with Bandit
make bandit

# Run Python tests
make test
```

### Docker deployment

```bash
# Build docker image
make docker-image

# Run web application and postgres db with docker-compose
make compose-up

# Check web application logs
docker logs whyemetl_server

# Check db logs
docker logs whyemetl_db

# Connect to db for interactive SQL
docker exec -it whyemetl_db bash
psql -U postgres

# Stop docker-compose
make compose-stop

# Run off docker-compose containers
make compose-down
```

### Data visualization

Metabase is run as part of the docker-compose deployment. Once you have run
`make compose-up`, Metabase server is available at `http://127.0.0.1:3000/`.

To use Metabase, you will to register yourself and indicate which database to
connect to:

- database: `db`
- database name: `postgres`
- database username: `postgres`
- database user password: `postgres`.

Once everything is set up, you should be able to navigate through Postgres data
by clicking on `Browse data`.

## Case study: Getting relevant information from open job positions data

### Dataset description

The dataset used here is stored under `data/dump.sql` that is an extract of a
Postgres database. The dump is automatically loaded when running both server and
database with docker-compose.

There are 2 tables:

- `jobs` that contains all the open job positions information. They comes from
  different companies. It is the fact table.
- `professions` that contains a list of professions that are associated to an
  open job positions. It is a dimension table.

From an analytics perspective, it is relevant to get statistics on the data such
as the distribution of professions by jobs or the repartition of number of open
positions by continent. As of today, the developped APIs focus on the latter use
case.

### Design considerations to get the number of jobs by continent

The data contains position information (latitude, longitude) of the offices
related to the open positions. Because the main `/jobsByContinent` relies on the
Continent information, it needed to be added in some way.

The first idea that came up was to integrate the PostGIS plugin that comes up
with multiple geospatial features on top of Postgres. While this idea sounded
promising, I was unsure the Continent column could be added easily. A Continent
can represented by a list of positions that would form a polygon. This
representation raised several questions such as:

- `how many points to take?`
- `who define what are the continent delimitations?`
- `where to get such data?`

Because both of these questions are non-trivial and more generally GIS is
complex, I decided not to take this approach. Moreover, in the future I might
want to deploy in a Cloud platform. Not all of their services contain an
out-of-the-box integration with PostGIS.

The second approach was to construct the Continent information dynamically for
every API call. While easier, I needed to find a way to get such information
given (latitude, longitude). Such information could be retrieved from the rich
Google Geolocation API. The downsides to this solution is that it would add a
high-adherence to an external service on which we do not have control and is
costly.

Both of these solutions seemed over-engineered and came with too many
drawbacks. Based on the assumption that job offices are very likely to be in
large capital cities, the final approach that was implemented is a mix of
static/dynamic information transformation:

- Approximate continents by Rectangle shapes roughly surrounding them
- Develop internal APIs to process the Continent information
- Add the information on existing information when starting the db (static
  transformations)
- Add the information dynamically when putting new data (not implemented).

See illustration below:

![Iteration 1](static/map.png)

This far more easier solution has good results:

- Only a few minutes is needed to process the data at rest when starting the database
- Most of the processing is done on the database side
- Developers can use SQL-like (filter, group by...) to get the right information
  instead of doing ad-hoc transformations in Python
- Total control on the transformation code
- Ease of maintenance using interfaces.

The major downside to this approach is the data quality issue. Considering
continents as rectangles (that can not be pivoted) is a raw approximation: the
rectangles will never match the boundaries of continents.

By adding the success_rate, it helped be to know if many rows were not labeled
or not. However, it does not solve the issue of low-quality rectangle
approximation. To improve the data quality, I would favor an iterative approach:

- Visualize a scatter-plotting of positions using Jupyter and Matplolib to check
  the rectangles
- Refine the rectangles positions and size.
- When rectangles is not precise enough, it might be interesting to try another shape.


ChangeLog
CHANGES
=======

* add code coverage
* add code coverage
* add circleci badge in readme
* improve circleci pipeline
* improve circleci pipeline
* fix circleci pipeline
* Remove workflow usage in ci
* Fix Python docker image version ci
* test ci
* Add .circleci/config.yml
* initial commit
* Initial commit


tests/conftest.py
import pytest

from whyemetl import create_app, db
from whyemetl.location import Position, Rectangle
from whyemetl.models import Jobs

#
# Fixture for location module testing
#


@pytest.fixture(scope="session")
def rectangular_europe():
    """ Raw approximation of Europe Area """
    upperleft = Position(58.950306, -17.455528)  # North-West of UK
    bottomright = Position(42.324367, 27.273977)  # South-East of Bulgaria
    return Rectangle("europe", upperleft, bottomright)


@pytest.fixture(scope="session")
def rectangular_america():
    """ Raw approximation of America Area """
    upperleft = Position(71.608584, -168.455893)  # Somewhere in Chukchi Sea
    bottomright = Position(23.437493, -52.142073)  # North-east or Porto Rico
    return Rectangle("america", upperleft, bottomright)


@pytest.fixture(scope="session")
def rectangular_asia():
    """ Raw approximation of America Area """
    upperleft = Position(71.654537, 37.787070)  # Somewhere in Barents Sea
    bottomright = Position(15.050491, 144.707127)  # Somewhere in the Philippine Sea
    return Rectangle("asia", upperleft, bottomright)


#
# Fixture for models module testing
#


@pytest.fixture(scope="function")
def new_job():
    return Jobs(
        job_reference="job1234",
        organization_reference="michel_organization",
        profession_id=1,
        office_latitude=48.866683,
        office_longitude=2.342687,
        office_continent="europe",
        candidates=3,
    )


#
# Fixture for functional/service testing
#


@pytest.fixture(scope="module")
def test_client():
    app = create_app()
    app.testing = True
    testing_client = app.test_client()

    with app.app_context():

        # Create the database and the database table
        db.create_all()

        # Organization 1
        # Europe
        # 2 different jobs with 2 different professions
        job1 = Jobs(
            job_reference="job1",
            organization_reference="michel_corp",
            profession_id=1,
            office_latitude=48.866683,
            office_longitude=2.342687,
            # continent not initialized (should be Europe)
            candidates=3,
        )

        job2 = Jobs(
            job_reference="job2",
            organization_reference="michel_corp",
            profession_id=2,
            office_latitude=48.866683,
            office_longitude=2.342687,
            # continent not initialized (should be Europe)
            candidates=1,
        )

        # Organization 2
        # Europe
        # 1 job with same profession of job1
        job3 = Jobs(
            job_reference="job3",
            organization_reference="toto_corp",
            profession_id=1,
            office_latitude=48.866683,
            office_longitude=2.342687,
            # continent not initialized (should be Europe)
            candidates=1,
        )

        # Organization 3
        # Asia
        # 2 jobs with 2 different professions. 1 is same than in Europe
        job4 = Jobs(
            job_reference="job4",
            organization_reference="beijing_corp",
            profession_id=1,
            office_latitude=39.880066,
            office_longitude=116.388496,
            # continent not initialized (should be Asia)
            candidates=1,
        )

        job5 = Jobs(
            job_reference="job5",
            organization_reference="beijing_corp",
            profession_id=4,
            office_latitude=39.880066,
            office_longitude=116.388496,
            # continent not initialized (should be Asia)
            candidates=1,
        )

        # Organization 4
        # Far south of Earth
        # 1 random job
        job6 = Jobs(
            job_reference="job6",
            organization_reference="nowhere_corp",
            profession_id=1,
            office_latitude=-66.123957,
            office_longitude=29.799831,
            # continent not initialized (should be OTHERS)
            candidates=1,
        )

        # Add mocked data in mocked db
        db.session.add(job1)
        db.session.add(job2)
        db.session.add(job3)
        db.session.add(job4)
        db.session.add(job5)
        db.session.add(job6)
        db.session.commit()

        yield testing_client

        db.drop_all()


tests/test_models.py
#
# Sanity test for Job model
#


def test_new_job(new_job):
    """
    GIVEN a new Job
    WHEN a Job is created
    THEN check the values
    """

    assert new_job.job_reference == "job1234"
    assert new_job.organization_reference == "michel_organization"
    assert new_job.profession_id == 1
    assert new_job.office_latitude == 48.866683
    assert new_job.office_longitude == 2.342687
    assert new_job.office_continent == "europe"
    assert new_job.candidates == 3


tests/test_service.py
"""
Functional testing at API level.
"""

import json


def test_hello(test_client):
    """
    GIVEN a Flask application
    WHEN the '/' page is requested (GET)
    THEN check the response is valid
    """

    response = test_client.get("/")
    assert response.status_code == 200
    assert b"Hello, World!" in response.data


def test_check_db_ctx(test_client):
    """
    GIVEN a Flask application AND a init db with mocked data
    WHEN the /checkDbConnection is requested (GET)
    THEN the response should contains the inserted first row
    """

    response = test_client.get("/checkDbConnection")
    data = json.loads(response.data)

    assert response.status_code == 200
    assert data["job_reference"] == "job1"
    assert data["organization_reference"] == "michel_corp"
    assert data["profession_id"] == 1
    assert data["office_latitude"] == 48.866683
    assert data["office_longitude"] == 2.342687
    assert data["office_continent"] is None
    assert data["candidates"] == 3


def test_consolidate_data(test_client):
    """
    GIVEN a Flask application AND a init db with mocked data
    WHEN the /consolidateData is requested (GET)
    AND /checkDbConnection is requested (GET)
    THEN the response should contains the inserted first row with updated continent
    """

    test_client.get("/consolidateData")
    response = test_client.get("/checkDbConnection")
    data = json.loads(response.data)

    assert response.status_code == 200
    assert data["office_continent"] == "EUROPE"


def test_get_jobs_by_continent(test_client):
    """
    GIVEN a Flask application AND a init db with mocked data
    WHEN the /consolidateData is requested (GET)
    AND /jobsByContinent is requested (POST) with 1 correct profession_id
    THEN the response should contains the right aggregated jobs by continent
    """

    test_client.get("/consolidateData")

    mimetype = "application/json"
    headers = {"Content-Type": mimetype}
    data = {"profession_ids": [1]}
    url = "/jobsByContinent"

    response = test_client.post(url, json=data, headers=headers)
    data = json.loads(response.data)

    assert response.status_code == 200
    assert data["ASIA"] == 1
    assert data["EUROPE"] == 4
    assert data["OTHERS"] == 1


def test_get_jobs_by_continent_multiple_profession_ids(test_client):
    """
    GIVEN a Flask application AND a init db with mocked data
    WHEN the /consolidateData is requested (GET)
    AND /jobsByContinent is requested (POST) with 2 correct profession_ids
    THEN the response should contains the right aggregated jobs by continent
    """

    test_client.get("/consolidateData")

    mimetype = "application/json"
    headers = {"Content-Type": mimetype}
    data = {"profession_ids": [1, 4]}
    url = "/jobsByContinent"

    response = test_client.post(url, json=data, headers=headers)
    data = json.loads(response.data)

    assert response.status_code == 200
    assert data["ASIA"] == 2
    assert data["EUROPE"] == 4
    assert data["OTHERS"] == 1


def test_get_jobs_by_continent_malformed_body(test_client):
    """
    GIVEN a Flask application AND a init db with mocked data
    WHEN the /consolidateData is requested (GET)
    AND /jobsByContinent is requested (POST) with wrong values
    THEN 400 error response should be returned
    """

    test_client.get("/consolidateData")

    mimetype = "application/json"
    headers = {"Content-Type": mimetype}
    url = "/jobsByContinent"

    # Negative value
    data = {"profession_ids": [-1]}
    response = test_client.post(url, json=data, headers=headers)
    assert response.status_code == 400

    # Wrong data type
    data = {"profession_ids": "Not a list"}
    response = test_client.post(url, json=data, headers=headers)
    assert response.status_code == 400

    # Empty list
    data = {"profession_ids": []}
    response = test_client.post(url, json=data, headers=headers)
    assert response.status_code == 400

    # Wrong key
    data = {"wrongkey": [1]}
    response = test_client.post(url, json=data, headers=headers)
    assert response.status_code == 400


def test_get_jobs_by_continent_not_found(test_client):
    """
    GIVEN a Flask application AND a init db with mocked data
    WHEN the /consolidateData is requested (GET)
    AND /jobsByContinent is requested (POST) with unknown profession_ids
    THEN 404 error response should be returned
    """

    test_client.get("/consolidateData")

    mimetype = "application/json"
    headers = {"Content-Type": mimetype}
    url = "/jobsByContinent"

    data = {"profession_ids": [999]}
    response = test_client.post(url, json=data, headers=headers)
    assert response.status_code == 404


whyemetl/views.py
"""
Contains web API routes logic.
"""

from flask import Blueprint, abort, request

from whyemetl import db, log
from whyemetl.repository import SQLJobsRepository
from whyemetl.service import WhyemetlDemoService

bp = Blueprint("/", __name__)


@bp.route("/")
def hello():
    return f"Hello, World!"


@bp.route("/checkDbConnection")
def check_db_ctx():
    service = WhyemetlDemoService(SQLJobsRepository(db))
    return service.check_db_ctx()


@bp.route("/consolidateData")
def consolidate_data():
    service = WhyemetlDemoService(SQLJobsRepository(db))
    return service.consolidate_data()


@bp.route("/jobsByContinent", methods=["POST"])
def get_jobs_by_continent():
    profession_ids = request.json.get("profession_ids")
    if malformed_body(profession_ids):
        log.info(f"Request body is malformed: {profession_ids}, returning 400 error")
        abort(400)
    try:
        service = WhyemetlDemoService(SQLJobsRepository(db))
        return service.get_jobs_by_continent(profession_ids)
    except Exception:
        log.info(f"No jobs found for {profession_ids}, returning 404 error")
        abort(404)


def malformed_body(body):
    if body is None:
        return True
    if not body:
        return True
    if not isinstance(body, list):
        return True
    if any(p < 0 for p in body):
        return True
    return False


install-packages.sh
#!/bin/bash

# Bash "strict mode", to help catch problems and bugs in the shell
# script. Every bash script you write should include this. See
# http://redsymbol.net/articles/unofficial-bash-strict-mode/ for
# details.
set -euo pipefail

# Tell apt-get we're never going to be able to give manual
# feedback:
export DEBIAN_FRONTEND=noninteractive

# Update the package listing, so we know what package exist:
apt-get update -y

# Install security updates:
apt-get -y upgrade

# Install a new package, without unnecessary recommended packages:
apt-get -y install --no-install-recommends \
    python3-pip \
    python3-dev \
    build-essential \
    vim \
    postgresql-client

# TODO
# Build WSGI Flask image instead of using Flask base server (dev)

# Delete cached files we don't need anymore:
apt-get clean
rm -rf /var/lib/apt/lists/*


whyemetl/service.py
"""
Contains business logic interfaces.
"""

from abc import ABC, abstractmethod

from whyemetl import america, asia, europe, log
from whyemetl.location import Position, get_continent


class WhyemetlService(ABC):
    """ Main business logic interfaces matching web APIs. """

    def __init__(self, repository):
        self.repository = repository

    @abstractmethod
    def check_db_ctx(self):
        """ Check connectivity with database. """
        pass

    @abstractmethod
    def consolidate_data(self):
        """ Consolidates Jobs data by adding Continent information. """
        pass

    @abstractmethod
    def get_jobs_by_continent(self):
        """
        Returns the aggregated jobs information by continent given
        the profession_ids.
        """
        pass


class WhyemetlDemoService(WhyemetlService):
    """ Demo implementation of WhyemetlService interface. """

    def __init__(self, repository):
        super().__init__(repository)

    def check_db_ctx(self):
        job = self.repository.get_first_job()

        return {
            "job_reference": job.job_reference,
            "organization_reference": job.organization_reference,
            "profession_id": job.profession_id,
            "office_latitude": job.office_latitude,
            "office_longitude": job.office_longitude,
            "office_continent": job.office_continent,
            "candidates": job.candidates,
        }

    def consolidate_data(self):
        count = self.repository.get_jobs_count()
        jobs = self.repository.get_jobs()

        log.info(f"{count} rows to process")
        log.info("Consolidating data with Continent information...")

        outlier_references = []
        outlier_count = 0

        # Update row by row while computing statistic on number of continents not found
        for job in jobs:
            position = Position(job.office_latitude, job.office_longitude)
            continent = get_continent(position, [europe, asia, america])

            if continent == "OTHERS":
                outlier_references.append(job.job_reference)
                outlier_count += 1
            job.office_continent = continent

        success_rate = 100 - (outlier_count / count * 100)
        log.info("Committing consolidated data to db...")
        self.repository.update_jobs()

        log.info("Consolidated data added to db with success")
        log.info(f"Could not locate Continent for {outlier_count} among {count} rows")
        log.info(f"Consolidation success rate is {success_rate}")

        return {
            "total_processed_rows": count,
            "total_failed_consolidation": outlier_count,
            "consolidation_success_rate": success_rate,
            "failed_rows": outlier_references,
        }

    def get_jobs_by_continent(self, profession_ids):
        results = self.repository.get_jobs_by_continent(profession_ids)
        if not results:
            raise Exception
        return {row.office_continent.upper(): row.total_candidates for row in results}


whyemetl/location.py
"""
Location contains utility classes to represents geopositions and continents.
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass


@dataclass
class Position:
    latitude: float
    longitude: float


class Continent(ABC):
    """ Continent interface that represents a geographical Continent area. """

    @abstractmethod
    def contains(self, position):
        """ Returns True if the given position is in the Continent area. """
        pass


class Rectangle(Continent):
    """A Continent area is approximated by a rectangle, defined by 2 positions:
    upperleft and bottomright"""

    def __init__(self, name, upperleft, bottomright):
        self.name = name
        self.upperleft = upperleft
        self.bottomright = bottomright

    def contains(self, position):
        """ Returns True if the given position is within the Continent. """

        if position.latitude is None or position.longitude is None:
            return False
        longitude_in_range = (
            self.upperleft.longitude <= position.longitude <= self.bottomright.longitude
        )
        latitude_in_range = (
            self.bottomright.latitude <= position.latitude <= self.upperleft.latitude
        )
        if longitude_in_range and latitude_in_range:
            return True
        return False


def get_continent(position, continents):
    """
    Checks in which continent the position is.
    Returns the continent name that contains it.
    """

    for continent in continents:
        if continent.contains(position):
            return continent.name
    return "OTHERS"


whyemetl/repository.py
"""
Contains interfaces for database comunication layer.
"""

from abc import ABC, abstractmethod

from sqlalchemy import func

from whyemetl.models import Jobs


class JobsRepository(ABC):
    """
    Repository interface to interact with Jobs model.
    """

    @abstractmethod
    def get_first_job(self):
        """ Returns the first row among jobs. """
        pass

    @abstractmethod
    def get_jobs(self):
        """ Returns all the jobs. """
        pass

    @abstractmethod
    def get_jobs_count(self):
        """ Returns total rows count. """
        pass

    @abstractmethod
    def update_jobs(self):
        """ Updates staged jobs to database. """
        pass

    @abstractmethod
    def get_jobs_by_continent(self, profession_ids):
        """
        Returns the aggregated jobs information by continent given
        the profession_ids.
        """
        pass


class SQLJobsRepository(JobsRepository):
    """
    SQL implementation of JobsRepository using SQLAlchemy.
    """

    def __init__(self, db):
        self.db = db

    def get_first_job(self):
        return self.db.session.query(Jobs).first()

    def get_jobs(self):
        return self.db.session.query(Jobs).all()

    def get_jobs_count(self):
        return self.db.session.query(Jobs).count()

    def update_jobs(self):
        self.db.session.commit()

    def get_jobs_by_continent(self, profession_ids):
        return (
            self.db.session.query(
                Jobs.office_continent,
                func.sum(Jobs.candidates).label("total_candidates"),
            )
            .filter(Jobs.profession_id.in_(profession_ids))
            .group_by(Jobs.office_continent)
            .all()
        )


whyemetl/__init__.py
import logging
import os

from flask import Flask
from flask_sqlalchemy import SQLAlchemy

from whyemetl.location import Position, Rectangle


def get_logger():
    """ Returns a well formatted logger with INFO level streaming out to stdout. """

    fmt = logging.Formatter(
        "[%(asctime)s] [%(filename)s:%(funcName)s:%(lineno)d] [%(levelname)s] - %(message)s"
    )
    logger = logging.getLogger(__name__)

    stream_handler = logging.StreamHandler()
    stream_handler.setFormatter(fmt)

    logger.addHandler(stream_handler)
    loglevel = logging.INFO

    logger.setLevel(loglevel)
    return logger


#
# Global variables
#

log = get_logger()
db = SQLAlchemy()


#
# Utility functions
#


def create_app():
    """ Returns a Flask application with database ORM and blueprint initialized. """

    from . import views

    app = Flask(__name__)
    app.register_blueprint(views.bp)
    db.init_app(app)
    return app


def db_url():
    """ Returns database URL for Postgres with docker-compose integration. """

    db_host = os.environ["PGHOST"]
    db_name = os.environ["PGDATABASE"]
    db_user = os.environ["PGUSER"]
    db_password = os.environ["PGPASSWORD"]
    return f"postgresql+psycopg2://{db_user}:{db_password}@{db_host}/{db_name}"


#
# Global geo-continent approximated by Rectangle
#


upperleft = Position(58.950306, -17.455528)  # North-West of UK
bottomright = Position(42.324367, 27.273977)  # South-East of Bulgaria
europe = Rectangle("EUROPE", upperleft, bottomright)

upperleft = Position(71.608584, -168.455893)  # Somewhere in Chukchi Sea
bottomright = Position(23.437493, -52.142073)  # North-east or Porto Rico
america = Rectangle("AMERICA", upperleft, bottomright)

upperleft = Position(71.654537, 37.787070)  # Somewhere in Barents Sea
bottomright = Position(15.050491, 144.707127)  # Somewhere in the Philippine Sea
asia = Rectangle("ASIA", upperleft, bottomright)
